src/software/gui/robot_diagnostics/widgets/robot_diagnostics_gui.cpp

```diff
@@ -33,24 +31,155 @@ void RobotDiagnosticsGUI::onChickerStateChanged(double chicker_power,
+                ->set_front_left_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_fl->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_front_right_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_fr->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_back_left_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_bl->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_back_right_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_br->text().toFloat());
+            break;
+        case WheelControlTab::DIRECT_VELOCITY:
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_velocity_control()
+                ->mutable_velocity()
+                ->set_x_component_meters(
+                    main_widget->lineEdit_direct_velocity_x->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_velocity_control()
+                ->mutable_velocity()
+                ->set_y_component_meters(
+                    main_widget->lineEdit_direct_velocity_y->text().toFloat());
+
+            auto angular_velocity_msg =
+                createAngularVelocityProto(AngularVelocity::fromDegrees(
+                    main_widget->lineEdit_direct_velocity_theta->text().toDouble()));
+            *(direct_control_primitive_msg->mutable_direct_control()
+                  ->mutable_direct_velocity_control()
+                  ->mutable_angular_velocity()) = *angular_velocity_msg;
+            break;
+    }
+
+    if (main_widget->buttonGroup_charge_state->checkedButton() ==
+        main_widget->radioButton_charge)
+    {
+        direct_control_primitive_msg->mutable_direct_control()->set_charge_mode(
+            TbotsProto::DirectControlPrimitive::ChargeMode::
+                DirectControlPrimitive_ChargeMode_CHARGE);
+    }
+    else if (main_widget->buttonGroup_charge_state->checkedButton() ==
+             main_widget->radioButton_discharge)
+    {
+        direct_control_primitive_msg->mutable_direct_control()->set_charge_mode(
+            TbotsProto::DirectControlPrimitive::ChargeMode::
+                DirectControlPrimitive_ChargeMode_DISCHARGE);
+    }
+    else if (main_widget->buttonGroup_charge_state->checkedButton() ==
+             main_widget->radioButton_float)
+    {
+        direct_control_primitive_msg->mutable_direct_control()->set_charge_mode(
+            TbotsProto::DirectControlPrimitive::ChargeMode::
+                DirectControlPrimitive_ChargeMode_FLOAT);
+    }
```

Comment:
can we refactor this to `direct_control_primitive_msg->mutable_direct_control()->set_charge_mode(getChargeMode())` and similarly with some of the other parts of this function so it's not so long?

Reply:
just to clarify, do you mean adding a member variable for ChargeMode?

Reply:
no, I mean making helper functions so that `createDirectControlPrimitiveFromUI` is easier to read